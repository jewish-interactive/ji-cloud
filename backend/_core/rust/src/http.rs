use std::net::{SocketAddr, TcpListener};

#[cfg(feature = "local")]
pub fn get_tcp_fd() -> Option<TcpListener> {
    listenfd::ListenFd::from_env().take_tcp_listener(0).unwrap()
}

#[cfg(not(feature = "local"))]
pub fn get_tcp_fd() -> Option<TcpListener> {
    // we don't have listenfd here.
    None
}

pub fn get_addr(default: u16) -> SocketAddr {
    let mut port = default;

    match std::env::var("PORT") {
        Ok(p) => {
            match p.parse::<u16>() {
                Ok(n) => {
                    port = n;
                }
                Err(_e) => {}
            };
        }
        Err(_e) => {}
    };

    ([0, 0, 0, 0], port).into()
}
