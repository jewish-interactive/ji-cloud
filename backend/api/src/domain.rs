
use shared::domain::category::{Category, CategoryId};
use uuid::Uuid;
use chrono::{DateTime, Utc};
use std::collections::HashMap;

#[derive(Debug)]
pub(crate) struct RawCategory {
    pub id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: String,
    pub index: i16,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub image_count: i64,
    pub jig_count: i64,
}

pub(crate) fn build_tree(mut categories: Vec<RawCategory>) -> Vec<Category> {
    let mut parent_to_id_index: HashMap<Option<Uuid>, Vec<usize>> = HashMap::new();

    for (idx, category) in categories.iter().enumerate() {
        parent_to_id_index
            .entry(category.parent_id)
            .and_modify(|it| it.push(idx))
            .or_insert_with(|| vec![idx]);
    }

    parent_to_id_index
        .values_mut()
        .for_each(|it| it.sort_unstable_by_key(|v| categories[*v].index));

    build_tree_recursive(&parent_to_id_index, &mut categories, None)
}

fn build_tree_recursive(
    parent_to_id_index: &HashMap<Option<Uuid>, Vec<usize>>,
    categories: &mut Vec<RawCategory>,
    seed_id: Option<Uuid>,
) -> Vec<Category> {
    let indecies = match parent_to_id_index.get(&seed_id) {
        Some(indecies) => indecies,
        None => return Vec::new(),
    };

    indecies
        .iter()
        .copied()
        .map(|category_index| {
            let children = build_tree_recursive(
                parent_to_id_index,
                categories,
                Some(categories[category_index].id),
            );

            let raw: &mut RawCategory = &mut categories[category_index];

            Category {
                id: CategoryId(raw.id),
                name: raw.name.clone(),
                created_at: raw.created_at,
                updated_at: raw.updated_at,
                image_count: raw.image_count as u64,
                jig_count: raw.jig_count as u64,
                children,
            }
        })
        .collect()
}
