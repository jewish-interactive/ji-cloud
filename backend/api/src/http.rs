mod auth;
mod cors;
mod endpoints;

use crate::{jwkkeys::JwkVerifier, s3};
use actix_service::Service;
use actix_web::dev::{MessageBody, ServiceRequest, ServiceResponse};
use config::JSON_BODY_LIMIT;
use core::{
    http::{get_addr, get_tcp_fd},
    settings::RuntimeSettings,
};
use futures::Future;
use s3::S3Client;
use sqlx::postgres::PgPool;
use std::sync::Arc;
use tokio::sync::RwLock;

fn log_ise<B: MessageBody, T>(
    req: ServiceRequest,
    srv: &mut T,
) -> impl Future<Output = actix_web::Result<T::Response>>
where
    T: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error>,
{
    let fut = srv.call(req);
    async {
        let res = fut.await?;

        if res.status() == 500 {
            if let Some(err) = res.response().extensions().get::<anyhow::Error>() {
                log::error!("ISE while responding to request: {:?}", err);
            }
        }

        Ok(res)
    }
}

#[actix_web::main]
pub async fn run(
    pool: PgPool,
    settings: RuntimeSettings,
    jwk_verifier: Arc<RwLock<JwkVerifier>>,
    s3: S3Client,
) -> anyhow::Result<()> {
    let local_insecure = settings.is_local();
    let api_port = settings.api_port;
    let server = actix_web::HttpServer::new(move || {
        actix_web::App::new()
            .data(pool.clone())
            .data(settings.clone())
            .data(s3.clone())
            .app_data(jwk_verifier.clone())
            .wrap(actix_web::middleware::Logger::default())
            .wrap_fn(log_ise)
            .wrap(cors::get(local_insecure).finish())
            .app_data(actix_web::web::JsonConfig::default().limit(JSON_BODY_LIMIT as usize))
            .configure(endpoints::user::configure)
            .configure(endpoints::category::configure)
            .configure(endpoints::image::configure)
    });

    // if listenfd doesn't take a TcpListener (i.e. we're not running via
    // the command above), we fall back to explicitly binding to a given
    // host:port.
    let server: _ = if let Some(l) = get_tcp_fd() {
        server.listen(l)?
    } else {
        server.bind(get_addr(api_port))?
    };

    server.run().await.unwrap();

    Ok(())
}
