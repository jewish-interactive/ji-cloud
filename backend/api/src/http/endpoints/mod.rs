pub mod admin;
pub mod category;
pub mod image;
pub mod jig;
pub mod meta;
pub mod module;
pub mod user;
