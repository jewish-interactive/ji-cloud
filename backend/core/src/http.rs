use std::net::{SocketAddr, TcpListener};

/// Creates a [`TcpListener`] via `listenfd` if possible.
///
/// [`TcpListener`]: https://doc.rust-lang.org/stable/std/net/struct.TcpListener.html
pub fn get_tcp_fd() -> Option<TcpListener> {
    #[cfg(feature = "listenfd")]
    let fd = listenfd::ListenFd::from_env().take_tcp_listener(0).unwrap();

    #[cfg(not(feature = "listenfd"))]
    let fd = None;

    fd
}

/// Get the port to run the server on.
pub fn get_addr(default: u16) -> SocketAddr {
    let port = std::env::var("PORT").map_or(default, |it| it.parse().unwrap_or(default));

    ([0, 0, 0, 0], port).into()
}
