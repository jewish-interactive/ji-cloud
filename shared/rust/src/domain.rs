pub mod auth;
pub mod category;
pub mod image;
pub mod user;
pub mod meta;

fn deserialize_optional_field<'de, T, D>(deserializer: D) -> Result<Option<Option<T>>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: serde::Deserialize<'de>,
{
    Ok(Some(serde::Deserialize::deserialize(deserializer)?))
}
