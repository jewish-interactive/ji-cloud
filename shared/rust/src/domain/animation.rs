//! Types for animations.

use chrono::{DateTime, Utc};
#[cfg(feature = "backend")]
use paperclip::actix::Apiv2Schema;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::Publish;

/// Animation Variants
#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", derive(paperclip::actix::Apiv2Schema))]
#[repr(i16)]
pub enum AnimationKind {
    /// Gif Animation
    Gif = 0,
    /// Spritesheet Animation
    Spritesheet = 1,
}

impl AnimationKind {
    /// returns `self` in a string representation.
    #[must_use]
    pub const fn to_str(self) -> &'static str {
        match self {
            Self::Gif => "gif",
            Self::Spritesheet => "spritesheet",
        }
    }
}

/// Wrapper type around [`Uuid`], represents the ID of an animation.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct AnimationId(pub Uuid);

/// Response for getting a single audio file.
#[derive(Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct AnimationResponse {
    /// The animation's metadata.
    pub metadata: AnimationMetadata,
}

/// Over the wire representation of an animation's metadata.
#[derive(Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct AnimationMetadata {
    /// The animation's ID.
    pub id: AnimationId,

    /// The name of the animation.
    pub name: String,

    /// The description of the animation.
    pub description: String,

    /// Is the animation premium?
    pub is_premium: bool,

    /// When the animation should be considered published (if at all).
    pub publish_at: Option<DateTime<Utc>>,

    /// What kind of animation this is.
    pub kind: AnimationKind,

    /// Should the animation loop?
    pub is_looping: bool,

    /// When the animation was originally created.
    pub created_at: DateTime<Utc>,

    /// When the animation was last updated.
    pub updated_at: Option<DateTime<Utc>>,
}

// todo: # errors doc section
/// Request to create a new animation.
#[derive(Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct AnimationCreateRequest {
    /// The name of the animation.
    pub name: String,

    /// The description of the animation.
    pub description: String,

    /// Is the animation premium?
    pub is_premium: bool,

    /// When to publish the animation.
    ///
    /// If [`Some`] publish the animation according to the `Publish`. Otherwise, don't publish it.
    pub publish_at: Option<Publish>,

    /// What kind of animation this is.
    pub variant: AnimationKind,

    /// Should the animation loop?
    pub is_looping: bool,
}

into_uuid![AnimationId];
