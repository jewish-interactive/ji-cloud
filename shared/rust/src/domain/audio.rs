//! Types for audio files.

#[cfg(feature = "backend")]
use paperclip::actix::Apiv2Schema;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Types for user audio library.
pub mod user {
    #[cfg(feature = "backend")]
    use paperclip::actix::Apiv2Schema;

    use serde::{Deserialize, Serialize};

    use super::AudioId;

    /// Response for listing.
    #[derive(Serialize, Deserialize, Debug)]
    #[cfg_attr(feature = "backend", derive(Apiv2Schema))]
    pub struct UserAudioListResponse {
        /// the audio files returned.
        pub audio_files: Vec<UserAudioResponse>,
    }

    /// Response for getting a single audio file.
    #[derive(Serialize, Deserialize, Debug)]
    #[cfg_attr(feature = "backend", derive(Apiv2Schema))]
    pub struct UserAudioResponse {
        /// The audio file's metadata.
        pub metadata: UserAudio,
    }

    /// Over the wire representation of an audio file's metadata.
    #[derive(Serialize, Deserialize, Debug)]
    #[cfg_attr(feature = "backend", derive(Apiv2Schema))]
    pub struct UserAudio {
        /// The audio file's ID.
        pub id: AudioId,
        // more fields to be added
    }
}

/// Wrapper type around [`Uuid`](Uuid), represents the ID of a image.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct AudioId(pub Uuid);

into_uuid![AudioId];
