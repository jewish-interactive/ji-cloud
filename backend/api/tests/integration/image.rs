use http::StatusCode;
use serde_json::json;
use shared::domain::{image::ImageId, CreateResponse};
use uuid::Uuid;

use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
};

async fn create(
    styles: &[Uuid],
    age_ranges: &[Uuid],
    affiliations: &[Uuid],
    categories: &[Uuid],
    tags: &[Uuid],
) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .login()
        .json(&json!({
            "name": "test",
            "description": "testest",
            "is_premium": false,
            "publish_at": (),
            "styles": styles,
            "age_ranges": age_ranges,
            "affiliations": affiliations,
            "categories": categories,
            "tags": tags,
            "kind": "Canvas",
        }))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: CreateResponse<ImageId> = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".id" => "[id]"});

    Ok(())
}

#[actix_rt::test]
async fn create_no_meta() -> anyhow::Result<()> {
    create(&[], &[], &[], &[], &[]).await
}

#[actix_rt::test]
async fn create_with_styles() -> anyhow::Result<()> {
    create(
        &[
            "6389eaa0-de76-11ea-b7ab-0399bcf84df2".parse()?,
            "6389ff7c-de76-11ea-b7ab-9b5661dd4f70".parse()?,
        ],
        &[],
        &[],
        &[],
        &[],
    )
    .await
}

#[actix_rt::test]
async fn create_with_meta() -> anyhow::Result<()> {
    create(
        &["6389eaa0-de76-11ea-b7ab-0399bcf84df2".parse()?],
        &["f3722790-de76-11ea-b7ab-77b45e9af3ef".parse()?],
        &["c0cd4446-de76-11ea-b7ab-93987e8aa112".parse()?],
        &[],
        &["5e72c62e-a3a4-11eb-96e7-c78c34eb32ee".parse()?],
    )
    .await
}

async fn create_error(kind: &str, id: &str) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .login()
        .json(&json!({
            "name": "test",
            "description": "testest",
            "is_premium": false,
            "publish_at": null,
            "styles": [],
            "age_ranges": [],
            "affiliations": [],
            "categories": [],
            "tags": [],
            "kind": "Canvas",
            kind: [id],
        }))
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::UNPROCESSABLE_ENTITY);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body);

    Ok(())
}

#[actix_rt::test]
async fn create_with_style_error() -> anyhow::Result<()> {
    create_error("styles", "6389eaa0-de76-11ea-b7ab-0399bcf84df2").await
}

#[actix_rt::test]
async fn create_with_affiliation_error() -> anyhow::Result<()> {
    create_error("affiliations", "6389eaa0-de76-11ea-b7ab-0399bcf84df2").await
}

#[actix_rt::test]
async fn create_with_age_range_error() -> anyhow::Result<()> {
    create_error("age_ranges", "6389eaa0-de76-11ea-b7ab-0399bcf84df2").await
}

#[actix_rt::test]
async fn create_with_category_error() -> anyhow::Result<()> {
    create_error("categories", "6389eaa0-de76-11ea-b7ab-0399bcf84df2").await
}

#[actix_rt::test]
async fn create_with_tags_error() -> anyhow::Result<()> {
    create_error("tags", "6389eaa0-de76-11ea-b7ab-0399bcf84df2").await
}

#[actix_rt::test]
async fn get_metadata() -> anyhow::Result<()> {
    let app = initialize_server(&[
        Fixture::User,
        Fixture::MetaKinds,
        Fixture::Image,
        Fixture::MetaImage,
    ])
    .await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".metadata.updated_at" => "[timestamp]"});

    Ok(())
}

// todo: search; missing algolia, s3
// todo: upload; missing algolia, s3
// todo: delete; missing algolia, s3
// todo: delete: edge case (never uploaded, should work even without s3), missing algolia

async fn update(req: &serde_json::Value) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::MetaKinds, Fixture::Image]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .json(req)
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".metadata.updated_at" => "[timestamp]"});

    Ok(())
}

#[actix_rt::test]
async fn update_empty() -> anyhow::Result<()> {
    update(&json!({})).await
}

#[actix_rt::test]
async fn update_is_premium() -> anyhow::Result<()> {
    update(&json!({"is_premium": true})).await
}

#[actix_rt::test]
async fn update_styles() -> anyhow::Result<()> {
    update(&json!({"styles": ["6389eaa0-de76-11ea-b7ab-0399bcf84df2", "6389ff7c-de76-11ea-b7ab-9b5661dd4f70"]})).await
}

#[actix_rt::test]
async fn update_tags() -> anyhow::Result<()> {
    update(&json!({"tags": ["591a2a64-a3a4-11eb-96e7-6bc0e819bc5f", "5b032222-a3a4-11eb-96e7-dbc5742f1640"]})).await
}

mod tag {
    use crate::{
        fixture::Fixture,
        helpers::{initialize_server, LoginExt},
    };
    use http::StatusCode;
    use shared::domain::image::tag::{ImageTagCreateRequest, ImageTagUpdateRequest};

    #[actix_rt::test]
    async fn create() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .post(&format!("http://0.0.0.0:{}/v1/image/tag/{}", port, 3))
            .json(&ImageTagCreateRequest {
                display_name: "test name".to_owned(),
            })
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::CREATED);

        let body: serde_json::Value = resp.json().await?;

        insta::assert_json_snapshot!(body, { ".id" => "[id]" });

        app.stop(false).await;

        Ok(())
    }

    #[actix_rt::test]
    async fn create_conflict() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .post(&format!("http://0.0.0.0:{}/v1/image/tag/{}", port, 0))
            .json(&ImageTagCreateRequest {
                display_name: "test name".to_owned(),
            })
            .login()
            .send()
            .await?;

        assert_eq!(resp.status(), StatusCode::CONFLICT);

        app.stop(false).await;

        Ok(())
    }

    #[actix_rt::test]
    async fn list() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        log::info!("making request");

        let resp = client
            .get(&format!("http://0.0.0.0:{}/v1/image/tag/all", port,))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::OK);

        let body: serde_json::Value = resp.json().await?;

        insta::assert_json_snapshot!(body, { ".**.id" => "[id]" });

        app.stop(false).await;

        Ok(())
    }

    async fn update(index: i16, req: ImageTagUpdateRequest) -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .patch(&format!("http://0.0.0.0:{}/v1/image/tag/{}", port, index,))
            .json(&req)
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::NO_CONTENT);

        let resp = client
            .get(&format!("http://0.0.0.0:{}/v1/image/tag/all", port,))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::OK);

        let body: serde_json::Value = resp.json().await?;

        insta::assert_json_snapshot!(body, { ".**.id" => "[id]" });

        app.stop(false).await;

        Ok(())
    }

    #[actix_rt::test]
    async fn update_no_index() -> anyhow::Result<()> {
        update(
            0,
            ImageTagUpdateRequest {
                display_name: Some("test".to_owned()),
                index: None,
            },
        )
        .await
    }

    #[actix_rt::test]
    async fn update_with_index() -> anyhow::Result<()> {
        update(
            1,
            ImageTagUpdateRequest {
                display_name: Some("test".to_owned()),
                index: Some(15),
            },
        )
        .await
    }

    #[actix_rt::test]
    async fn update_none() -> anyhow::Result<()> {
        update(
            1,
            ImageTagUpdateRequest {
                display_name: None,
                index: None,
            },
        )
        .await
    }

    #[actix_rt::test]
    async fn update_only_index() -> anyhow::Result<()> {
        update(
            1,
            ImageTagUpdateRequest {
                display_name: None,
                index: Some(3),
            },
        )
        .await
    }

    #[actix_rt::test]
    async fn update_conflict() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .patch(&format!("http://0.0.0.0:{}/v1/image/tag/{}", port, 1))
            .json(&ImageTagUpdateRequest {
                display_name: None,
                index: Some(0),
            })
            .login()
            .send()
            .await?;

        assert_eq!(resp.status(), StatusCode::CONFLICT);

        let resp = client
            .get(&format!("http://0.0.0.0:{}/v1/image/tag/all", port,))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::OK);

        let body: serde_json::Value = resp.json().await?;

        insta::assert_json_snapshot!(body, { ".**.id" => "[id]" });

        app.stop(false).await;

        Ok(())
    }

    #[actix_rt::test]
    async fn delete() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .delete(&format!("http://0.0.0.0:{}/v1/image/tag/2", port,))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::NO_CONTENT);

        app.stop(false).await;

        Ok(())
    }
}

mod recent {
    use http::StatusCode;
    use serde_json::json;

    use crate::{
        fixture::Fixture,
        helpers::{initialize_server, LoginExt},
    };
    use shared::domain::image::recent::UserRecentImageListRequest;

    #[actix_rt::test]
    async fn create() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .post(&format!("http://0.0.0.0:{}/v1/user/me/recent/image", port))
            .json(&json!({
                "id": "3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
                "library": "User",
            }))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::CREATED);

        let body: serde_json::Value = resp.json().await?;

        app.stop(false).await;

        insta::assert_json_snapshot!(body, {".last_used" => "[timestamp]"});

        Ok(())
    }

    #[actix_rt::test]
    async fn create_conflict() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .post(&format!("http://0.0.0.0:{}/v1/user/me/recent/image", port))
            .json(&json!({
                //"id": "3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
                "id": "8cca6f3a-c4bb-11eb-8edf-13c75672da8f",
                "library": "User",
            }))
            .login()
            .send()
            .await?;

        assert_eq!(resp.status(), StatusCode::CONFLICT);

        let body: serde_json::Value = resp.json().await?;

        app.stop(false).await;

        insta::assert_json_snapshot!(body, {".last_used" => "[timestamp]"});

        Ok(())
    }

    #[actix_rt::test]
    async fn list_no_limit() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .get(&format!("http://0.0.0.0:{}/v1/user/me/recent/image", port))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::OK);

        let body: serde_json::Value = resp.json().await?;

        app.stop(false).await;

        insta::assert_json_snapshot!(body);

        Ok(())
    }

    #[actix_rt::test]
    async fn list_with_limit() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .get(&format!("http://0.0.0.0:{}/v1/user/me/recent/image", port))
            .login()
            .query(&UserRecentImageListRequest { limit: 3 })
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::OK);

        let body: serde_json::Value = resp.json().await?;

        app.stop(false).await;

        insta::assert_json_snapshot!(body);

        Ok(())
    }

    #[actix_rt::test]
    async fn update() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .patch(&format!(
                "http://0.0.0.0:{}/v1/user/me/recent/image/{}",
                port, "8cca719c-c4bb-11eb-8edf-f7accb638a15"
            ))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::NO_CONTENT);

        let resp = client
            .get(&format!("http://0.0.0.0:{}/v1/user/me/recent/image", port))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::OK);

        let body: serde_json::Value = resp.json().await?;

        app.stop(false).await;

        insta::assert_json_snapshot!(body, {".**.last_used" => "[timestamp]"});

        Ok(())
    }

    #[actix_rt::test]
    async fn delete() -> anyhow::Result<()> {
        let app = initialize_server(&[Fixture::User, Fixture::Image]).await;

        let port = app.port();

        let client = reqwest::Client::new();

        let resp = client
            .delete(&format!(
                "http://0.0.0.0:{}/v1/user/me/recent/image/{}",
                port, "8cca719c-c4bb-11eb-8edf-f7accb638a15"
            ))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::NO_CONTENT);

        let resp = client
            .get(&format!("http://0.0.0.0:{}/v1/user/me/recent/image", port))
            .login()
            .send()
            .await?
            .error_for_status()?;

        assert_eq!(resp.status(), StatusCode::OK);

        let body: serde_json::Value = resp.json().await?;

        app.stop(false).await;

        insta::assert_json_snapshot!(body); //, {".**.last_used" => "[timestamp]"});

        Ok(())
    }
}
