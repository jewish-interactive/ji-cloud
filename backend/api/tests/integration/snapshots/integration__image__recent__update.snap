---
source: tests/integration/image.rs
expression: body

---
{
  "images": [
    {
      "id": "8cca719c-c4bb-11eb-8edf-f7accb638a15",
      "library": "User",
      "last_used": "[timestamp]"
    },
    {
      "id": "8cca6f3a-c4bb-11eb-8edf-13c75672da8f",
      "library": "User",
      "last_used": "[timestamp]"
    },
    {
      "id": "8cca720a-c4bb-11eb-8edf-63da1d86939c",
      "library": "User",
      "last_used": "[timestamp]"
    },
    {
      "id": "8cca7124-c4bb-11eb-8edf-7b42383ed8f5",
      "library": "User",
      "last_used": "[timestamp]"
    }
  ]
}
